stablehlo/reference/Ops.cpp

```diff
@@ -102,6 +102,81 @@ Tensor evalSliceOp(const Tensor &operand, const Sizes &startIndices,
                      inferredTypes[0].cast<ShapedType>());
 }
 
+static Tensor computeSum(const Tensor &operand, const Tensor &initValue,
+                         const Axes &dimensions, ShapedType resultType) {
+  Tensor result(resultType);
+  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
+       ++resultIt)
+    result.set(*resultIt, initValue.get({}));
+
+  for (auto operandIt = operand.index_begin(); operandIt != operand.index_end();
+       ++operandIt) {
+    Index resultIndex = *operandIt;
+    for (auto dim : llvm::reverse(dimensions))
+      resultIndex.erase(resultIndex.begin() + dim);
+
+    Tensor currVal(initValue.getType());
+    currVal.set({}, result.get(resultIndex));
+
+    Tensor nextValToAdd(initValue.getType());
+    nextValToAdd.set({}, operand.get(*operandIt));
+
+    auto addResult = evalAddOp(currVal, nextValToAdd, currVal.getType());
+    result.set(resultIndex, addResult.get({}));
+  }
+  return result;
+}
```

Comment:
Since we're partially evaluating the implementation of `evalReduceOp` anyway, we might as well partially evaluate it further and inline the call to evalAddOp:

```C++
static Tensor computeSum(const Tensor &input, const Tensor &initValue,
                         const Axes &dimensions, ShapedType resultType) {
  Tensor result(resultType);
  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
       ++resultIt)
    result.set(*resultIt, initValue.get({}));

  for (auto inputIt = input.index_begin(); inputIt != input.index_end();
       ++inputIt) {
    Index resultIndex;
    for (auto [inputAxis, inputIndexElement] : llvm::enumerate(*inputIt)) {
      if (llvm::is_contained(dimensions, inputAxis)) continue;
      resultIndex.push_back(inputIndexElement);
    }
    result.set(resultIndex, result.get(resultIndex) + input.get(*inputIt));
  }
  return result;
}
```